fenicsadapter/fenicsadapter.py

```diff
@@ -159,46 +164,86 @@ def create_coupling_neumann_boundary_condition(self, test_functions):
         self._write_data = self.convert_fenics_to_precice(write_function, self._mesh_fenics, self._coupling_subdomain)
+
+        # communication
         self._interface.writeBlockScalarData(self._write_data_id, self._n_vertices, self._vertex_ids, self._write_data)
         self._interface.advance(dt)
         self._interface.readBlockScalarData(self._read_data_id, self._n_vertices, self._vertex_ids, self._read_data)
+
+        # update boundary condition with read data
         self._coupling_bc_expression.update_boundary_data(self._read_data, x_vert, y_vert)
+
+        success = False
+
+        # checkpointing
         if self._interface.isActionRequired(PySolverInterface.PyActionReadIterationCheckpoint()):
-            # We do not require proper checkpointing for the FEniCS examples we are currently dealing with.
-            # do nothing
+            # continue FEniCS computation from checkpoint
+            # todo we might want to put reading the checkpoint into a function (duplicate code. compare to below)
+            u_n.assign(self._u_cp)  # set u_n to value of checkpoint
+            t = self._t_cp
+            n = self._n_cp
             self._interface.fulfilledAction(PySolverInterface.PyActionReadIterationCheckpoint())
-            return False
-        else:
-            return True
 
-    def initialize(self, coupling_subdomain, mesh, read_field, write_field):
+        if self._interface.isActionRequired(PySolverInterface.PyActionWriteIterationCheckpoint()):
+            # continue FEniCS computation with u_np1
+            # update checkpoint
+            self._u_cp.assign(u_np1)
+            assert (np.isclose(t + dt, self._t_cp + dt))
+            self._t_cp = t + dt
+            assert (np.isclose(n + 1, self._n_cp + 1))
+            self._n_cp = n + 1
+            # todo we might want to put reading the checkpoint into a function (duplicate code. compare to above)
+            u_n.assign(self._u_cp)  # set u_n to value of (updated)checkpoint
+            t = self._t_cp
+            n = self._n_cp
+            self._interface.fulfilledAction(PySolverInterface.PyActionWriteIterationCheckpoint())
+            success = True
+
+        return t, n, success
+
+    def initialize(self, coupling_subdomain, mesh, read_field, write_field, u_n, t=0, n=0):
         self.set_coupling_mesh(mesh, coupling_subdomain)
         self.set_read_field(read_field)
         self.set_write_field(write_field)
         self._precice_tau = self._interface.initialize()
 
         if self._interface.isActionRequired(PySolverInterface.PyActionWriteInitialData()):
-            self._interface.writeBlockScalarData(self._read_data_id, self._n_vertices, self._vertex_ids, self._read_data)
+            self._interface.writeBlockScalarData(self._write_data_id, self._n_vertices, self._vertex_ids, self._write_data)
             self._interface.fulfilledAction(PySolverInterface.PyActionWriteInitialData())
 
         self._interface.initializeData()
```

Comment:
better to put this also into the if-statement above, though not completely wrong (i just also updated the adapter example)

Reply:
This means I will just intend https://github.com/precice/fenics-adapter/pull/19#discussion-diff-248446873R234 by one tab. Easy fix :)

Reply:
ah, mistake. don't do that. my fault. leave the `initializaData()` out of the if-case